lib/semantics/resolve-names.cc

```diff
@@ -2960,29 +2998,64 @@ bool DeclarationVisitor::Pre(const parser::StructureConstructor &x) {
   return false;
 }
 
-Symbol *DeclarationVisitor::DeclareConstructEntity(const parser::Name &name) {
+Symbol *DeclarationVisitor::DeclareLocalEntity(const parser::Name &name) {
   auto *prev{FindSymbol(name)};
-  if (prev) {
-    if (prev->owner().kind() == Scope::Kind::Forall ||
-        prev->owner() == currScope()) {
+  bool implicit{false};
+  if (prev == nullptr) {
+    // Declare the name as an object in the enclosing scope so that
+    // the name can't be repurposed there later as something else.
+    prev = &MakeSymbol(InclusiveScope(), name.source, Attrs{});
+    ApplyImplicitRules(*prev);
+    implicit = true;
+  }
+  if (!ConvertToObjectEntity(*prev)) {
+    Say2(name, "Locality attribute not allowed on '%s'"_err_en_US, *prev,
```

Comment:
Is there a test for this case?

Reply:
Not that I can find.  I'll add one to this PR.